File: plugins/robots/robotsGenerator/nxtOSEK/generators/simpleElements/waitForColorBlockGenerator.cpp
#include "waitForColorBlockGenerator.h"
#include "../../nxtOSEKRobotGenerator.h"
#include "../listsmartline.h"

using namespace robots::generator;

WaitForColorBlockGenerator::WaitForColorBlockGenerator()
{
}

void WaitForColorBlockGenerator::addInitAndTerminateCode(NxtOSEKRobotGenerator *nxtGen,
		QString const &port, QString const &colorNxtType, qReal::Id const elementId)
{
	QString const partInitCode = "ecrobot_init_nxtcolorsensor(" + port;
	QString const initCode = "ecrobot_init_nxtcolorsensor(" + port + "," + colorNxtType + ");";

	if (!ListSmartLine::isContainsPart(nxtGen->initCode(), partInitCode)) {
		QString const terminateCode = "ecrobot_init_nxtcolorsensor(" + port + "," + colorNxtType + ");";
		nxtGen->initCode().append(SmartLine(initCode, elementId));
		nxtGen->terminateCode().append(SmartLine(terminateCode, elementId));
	}
}

QList<SmartLine> WaitForColorBlockGenerator::convertElementIntoDirectCommand(NxtOSEKRobotGenerator *nxtGen
		, qReal::Id const elementId, qReal::Id const logicElementId)
{
	QList<SmartLine> result;
	int const port = nxtGen->api()->stringProperty(logicElementId, "Port").toInt();
	QByteArray const colorStr = nxtGen->api()->stringProperty(logicElementId, "Color").toUtf8();

	QString colorNxtType;

	if (colorStr == "Красный") {
		colorNxtType = "NXT_COLOR_RED";
	} else if (colorStr == "Зелёный") {
		colorNxtType = "NXT_COLOR_GREEN";
	} else if (colorStr == "Синий") {
		colorNxtType = "NXT_COLOR_BLUE";
	} else if (colorStr == "Чёрный") {
		colorNxtType = "NXT_COLOR_BLACK";
	} else if (colorStr == "Жёлтый") {
		colorNxtType = "NXT_COLOR_YELLOW";
	} else if (colorStr == "Белый") {
		colorNxtType = "NXT_COLOR_WHITE";
	}

	if (!colorNxtType.isEmpty()) {
		QString portStr = QString::number(port);
		result.append(SmartLine("while (ecrobot_get_nxtcolorsensor_id(NXT_PORT_S" + portStr
				+ ") != " + colorNxtType + ")", elementId));
		result.append(SmartLine("{", elementId));
		result.append(SmartLine("}", elementId));
		addInitAndTerminateCode(nxtGen, portStr, colorNxtType, elementId);
	}

	return result;
}

File: plugins/visualInterpreter/examples/robotsCodeGeneration/reactionsStorage/WaitForColorBlockGenerator.py
# Application condition
waitFor.id == max_used_id and not cur_node_is_processed

# Reaction
port = "NXT_PORT_S" + waitFor.Port

color_nxt_type = ""
color_str = waitFor.Color

if color_str == "Красный":
  color_nxt_type = "NXT_COLOR_RED"
elif color_str == "Зелёный":
  color_nxt_type = "NXT_COLOR_GREEN"
elif color_str == "Синий":
  color_nxt_type = "NXT_COLOR_BLUE"
elif color_str == "Чёрный":
  color_nxt_type = "NXT_COLOR_BLACK"
elif color_str == "Жёлтый":
  color_nxt_type = "NXT_COLOR_YELLOW"
elif color_str == "Белый":
  color_nxt_type = "NXT_COLOR_WHITE"

wait_for_color_block_code = "while (ecrobot_get_nxtcolorsensor_id(" + port + ") != " + color_nxt_type + ") {}\n"
wait_init_code = "ecrobot_init_nxtcolorsensor(" + port + ", " + color_nxt_type + ");\n"
wait_terminate_code = "ecrobot_init_nxtcolorsensor(" + port + ", " + color_nxt_type + ");\n"

if wait_init_code not in init_code:
  init_code.append(wait_init_code)
  terminate_code.append(wait_terminate_code)

code.append([wait_for_color_block_code])
id_to_pos_in_code[waitFor.id] = len(code) - 1

cur_node_is_processed = True
